ambry-store/src/main/java/com.github.ambry.store/StatsManager.java

```diff
@@ -0,0 +1,186 @@
+   * second element is the number of {@link Store}s that were skipped (either unreachable or an error has occurred).
+   */
+  Pair<StatsSnapshot, List<String>> collectAndAggregate(Set<PartitionId> partitionIds) {
+    StatsSnapshot aggregatedStatsSnapshot = new StatsSnapshot(0L, null);
+    List<String> unreachableStores = new ArrayList<>();
+    for (PartitionId partitionId : partitionIds) {
+      Store store = storageManager.getStore(partitionId);
+      if (store != null) {
+        try {
+          StatsSnapshot statsSnapshot = store.getStoreStats().getStatsSnapshot();
+          aggregate(aggregatedStatsSnapshot, statsSnapshot);
+        } catch (StoreException e) {
+          logger.error("Store exception thrown when getting stats for partitionId: " + partitionId.toString(), e);
+          unreachableStores.add(partitionId.toString());
+        }
+      } else {
+        unreachableStores.add(partitionId.toString());
+      }
+    }
+    return new Pair<>(aggregatedStatsSnapshot, unreachableStores);
+  }
+
+  /**
+   * Performs recursive aggregation of two {@link StatsSnapshot} and stores the result in the first one.
+   * @param baseSnapshot one of the addends and where the result will be
+   * @param newSnapshot the other addend to be added into the first {@link StatsSnapshot}
+   */
+  private void aggregate(StatsSnapshot baseSnapshot, StatsSnapshot newSnapshot) {
+    baseSnapshot.setValue(baseSnapshot.getValue() + newSnapshot.getValue());
+    if (baseSnapshot.getSubtree() == null) {
+      baseSnapshot.setSubtree(newSnapshot.getSubtree());
+    } else if (newSnapshot.getSubtree() != null) {
+      for (Map.Entry<String, StatsSnapshot> entry : newSnapshot.getSubtree().entrySet()) {
+        if (!baseSnapshot.getSubtree().containsKey(entry.getKey())) {
+          baseSnapshot.getSubtree().put(entry.getKey(), new StatsSnapshot(0L, null));
+        }
+        aggregate(baseSnapshot.getSubtree().get(entry.getKey()), entry.getValue());
+      }
+    }
+  }
+
+  /**
+   * Runnable class that collects, aggregate and publish stats via methods in StatsManager.
+   */
+  private class StatsAggregator implements Runnable {
+
+    @Override
+    public void run() {
+      Set<PartitionId> partitionIds = storageManager.getPartitionIds();
+      Pair<StatsSnapshot, List<String>> result = collectAndAggregate(partitionIds);
+      List<String> unreachableStores = result.getSecond();
+      StatsHeader statsHeader =
+          new StatsHeader(Description.QUOTA, SystemTime.getInstance().milliseconds(), partitionIds.size(),
+              partitionIds.size() - unreachableStores.size(), unreachableStores);
+      try {
+        publish(new StatsWrapper(statsHeader, result.getFirst()));
+        logger.info("Stats snapshot published to " + statsOutputFile.getAbsolutePath());
+      } catch (IOException e) {
+        logger.error("IOException when publishing stats to " + statsOutputFile.getAbsolutePath(), e);
```

Comment:
metrics will be useful wherever errors are not propagated and just swallowed. Also any other useful metrics (how long things took etc)